common/pkg/capi/clustertopology/handlers/lifecycle/parallel.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2024 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package lifecycle
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/samber/lo"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+func runHooksInParallel[T runtimehooksv1.RequestObject, U runtimehooksv1.ResponseObject](
```

Comment:
One downside of this is that the logs are interleaved from different hooks. Any other potential downside to this?

Reply:
I can't think of any.

We can improve the logging by adding hook names to the log output to make the interleaved logs more readable, or rather to be able to trace through the individual hooks by filtering the logs. Generally the lifecycle hooks have few logs so it shouldn't cause any issue for now though.